File: python/class_hierarchy.py
"""
Python test file for class hierarchy changes.

This file demonstrates class inheritance structures 
to test DiffScope's ability to detect changes in class hierarchies.
"""

from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional, Union, Set


class Entity:
    """Base class for all entities in the system."""
    
    def __init__(self, id: str, name: str):
        self.id = id
        self.name = name
        self.tags = set()
        self.created_at = None  # New field
    
    def add_tag(self, tag: str) -> None:
        """Add a tag to this entity."""
        self.tags.add(tag)
        
    def remove_tag(self, tag: str) -> bool:
        """Remove a tag from this entity if it exists."""
        if tag in self.tags:
            self.tags.remove(tag)
            return True
        return False
    
    def has_tag(self, tag: str) -> bool:
        """Check if this entity has the specified tag."""
        return tag in self.tags
    
    def __str__(self) -> str:
        return f"{self.__class__.__name__}(id={self.id}, name={self.name})"
    
    def __repr__(self) -> str:
        return self.__str__()


# New interface
class Searchable(ABC):
    """Interface for objects that can be searched."""
    
    @abstractmethod
    def get_search_terms(self) -> List[str]:
        """Get search terms for this object."""
        pass


class Printable(ABC):
    """Interface for objects that can be printed."""
    
    @abstractmethod
    def to_printable_format(self) -> str:
        """Convert to a printable format."""
        pass
    
    # New method
    @abstractmethod
    def to_html(self) -> str:
        """Convert to HTML format."""
        pass


class Named(ABC):
    """Interface for objects that have a name attribute."""
    
    @abstractmethod
    def get_display_name(self) -> str:
        """Get the display name for this object."""
        pass


# Modified inheritance
class User(Entity, Named, Searchable):  # Added Searchable
    """Represents a user in the system."""
    
    def __init__(self, id: str, name: str, email: str, role: str = "standard"):
        super().__init__(id, name)
        self.email = email
        self.role = role
        self.permissions = set()
        self.last_login = None  # New field
        
    def add_permission(self, permission: str) -> None:
        """Add a permission to this user."""
        self.permissions.add(permission)
        
    def has_permission(self, permission: str) -> bool:
        """Check if this user has the specified permission."""
        return permission in self.permissions
    
    def get_display_name(self) -> str:
        """Get the display name for this user."""
        return f"{self.name} ({self.email})"
    
    # Implement Searchable
    def get_search_terms(self) -> List[str]:
        """Get search terms for this user."""
        return [self.id, self.name, self.email, self.role]


# New base class for resources
class BaseResource(Entity):
    """Base class for all resources."""
    
    def __init__(self, id: str, name: str, owner_id: str):
        super().__init__(id, name)
        self.owner_id = owner_id
        self.data = {}
    
    def update_data(self, key: str, value: Any) -> None:
        """Update a data field for this resource."""
        self.data[key] = value
        
    def get_data(self, key: str, default: Any = None) -> Any:
        """Get a data field from this resource."""
        return self.data.get(key, default)


# Modified inheritance
class Resource(BaseResource, Named, Printable, Searchable):  # Changed to extend BaseResource, added Searchable
    """Represents a resource in the system."""
    
    def __init__(self, id: str, name: str, owner_id: str, resource_type: str):
        super().__init__(id, name, owner_id)
        self.resource_type = resource_type
        
    def get_display_name(self) -> str:
        """Get the display name for this resource."""
        return f"{self.name} ({self.resource_type})"
    
    def to_printable_format(self) -> str:
        """Convert to a printable format."""
        return f"Resource: {self.name}\nType: {self.resource_type}\nOwner: {self.owner_id}\nData: {self.data}"
    
    # Implement abstract method
    def to_html(self) -> str:
        """Convert to HTML format."""
        return f"<div class='resource'><h2>{self.name}</h2><p>Type: {self.resource_type}</p></div>"
    
    # Implement Searchable
    def get_search_terms(self) -> List[str]:
        """Get search terms for this resource."""
        return [self.id, self.name, self.resource_type]


class Document(Resource):
    """Represents a document resource."""
    
    def __init__(self, id: str, name: str, owner_id: str, content: str = ""):
        super().__init__(id, name, owner_id, "document")
        self.content = content
        self.version = 1
        
    def update_content(self, new_content: str) -> None:
        """Update the document content."""
        self.content = new_content
        self.version += 1
        
    def to_printable_format(self) -> str:
        """Convert to a printable format."""
        base_format = super().to_printable_format()
        return f"{base_format}\nVersion: {self.version}\nContent: {self.content[:50]}..."
    
    # Override the HTML format
    def to_html(self) -> str:
        """Convert to HTML format."""
        return f"""
        <div class='document'>
            <h2>{self.name}</h2>
            <p>Version: {self.version}</p>
            <div class='content'>{self.content[:100]}...</div>
        </div>
        """
    
    # Override search terms
    def get_search_terms(self) -> List[str]:
        """Get search terms for this document."""
        terms = super().get_search_terms()
        # Add content as search term
        terms.append(self.content)
        return terms


class Image(Resource):
    """Represents an image resource."""
    
    def __init__(self, id: str, name: str, owner_id: str, width: int, height: int, format: str = "png"):
        super().__init__(id, name, owner_id, "image")
        self.width = width
        self.height = height
        self.format = format
        
    def to_printable_format(self) -> str:
        """Convert to a printable format."""
        base_format = super().to_printable_format()
        return f"{base_format}\nDimensions: {self.width}x{self.height}\nFormat: {self.format}"
    
    # Override the HTML format
    def to_html(self) -> str:
        """Convert to HTML format."""
        return f"""
        <div class='image'>
            <h2>{self.name}</h2>
            <p>Dimensions: {self.width}x{self.height}</p>
            <p>Format: {self.format}</p>
            <img src="placeholder.jpg" alt="{self.name}" />
        </div>
        """


# New class
class GuestUser(User):
    """Represents a guest user with limited permissions."""
    
    def __init__(self, id: str, name: str = "Guest"):
        super().__init__(id, name, f"guest_{id}@example.com", role="guest")
    
    def get_display_name(self) -> str:
        """Get the display name for this guest."""
        return f"Guest: {self.name}"


# Modified class
class AdminUser(User):
    """Represents an admin user with extended permissions."""
    
    def __init__(self, id: str, name: str, email: str, admin_level: int = 1):  # Added admin_level
        super().__init__(id, name, email, role="admin")
        self.admin_level = admin_level  # New field
        # Default admin permissions
        self.add_permission("create_user")
        self.add_permission("delete_user")
        self.add_permission("edit_resource")
        
        # Higher level admins get more permissions
        if admin_level >= 2:
            self.add_permission("manage_roles")
            self.add_permission("view_logs")
        
        if admin_level >= 3:
            self.add_permission("system_config")
        
    def can_manage_users(self) -> bool:
        """Check if this admin can manage users."""
        return self.has_permission("create_user") and self.has_permission("delete_user")
        
    def get_display_name(self) -> str:
        """Get the display name for this admin."""
        return f"Admin(L{self.admin_level}): {self.name} ({self.email})"


# Demo usage
if __name__ == "__main__":
    # Create an admin user
    admin = AdminUser("admin1", "Admin User", "admin@example.com", admin_level=2)
    admin.add_tag("administrator")
    
    # Create a standard user
    user = User("user1", "Standard User", "user@example.com")
    user.add_tag("beta-tester")
    
    # Create a guest user
    guest = GuestUser("guest1")
    guest.add_tag("temporary")
    
    # Create a document
    doc = Document("doc1", "Important Document", "user1", "This is an important document")
    doc.add_tag("important")
    
    # Create an image
    img = Image("img1", "Profile Photo", "user1", 1200, 800)
    img.add_tag("profile")
    
    # Demonstrate hierarchy
    entities = [admin, user, guest, doc, img]
    for entity in entities:
        print(f"\n{entity}")
        
        if isinstance(entity, Named):
            print(f"Display name: {entity.get_display_name()}")
            
        if isinstance(entity, Printable):
            print(f"Printable format:\n{entity.to_printable_format()}")
            
        if isinstance(entity, Searchable):
            print(f"Search terms: {', '.join(entity.get_search_terms())}")
            
        print(f"Tags: {', '.join(entity.tags)}") 

File: javascript/class_hierarchy.js
/**
 * JavaScript test file for class hierarchy changes.
 * 
 * This file demonstrates class inheritance structures 
 * to test DiffScope's ability to detect changes in class hierarchies.
 */

/**
 * Base class for all entities in the system.
 */
class Entity {
  /**
   * Create a new entity.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   */
  constructor(id, name) {
    this.id = id;
    this.name = name;
    this.tags = new Set();
  }
  
  /**
   * Add a tag to this entity.
   * @param {string} tag - Tag to add
   */
  addTag(tag) {
    this.tags.add(tag);
  }
  
  /**
   * Remove a tag from this entity if it exists.
   * @param {string} tag - Tag to remove
   * @returns {boolean} True if tag was removed, false if it didn't exist
   */
  removeTag(tag) {
    if (this.tags.has(tag)) {
      this.tags.delete(tag);
      return true;
    }
    return false;
  }
  
  /**
   * Check if this entity has the specified tag.
   * @param {string} tag - Tag to check
   * @returns {boolean} True if entity has the tag
   */
  hasTag(tag) {
    return this.tags.has(tag);
  }
  
  /**
   * String representation of this entity.
   * @returns {string} String representation
   */
  toString() {
    return `${this.constructor.name}(id=${this.id}, name=${this.name})`;
  }
}

/**
 * Interface for objects that can be printed.
 */
class Printable {
  /**
   * Convert to a printable format.
   * @abstract
   * @returns {string} Printable representation
   */
  toPrintableFormat() {
    throw new Error('Method toPrintableFormat() must be implemented');
  }
}

/**
 * Interface for objects that have a name attribute.
 */
class Named {
  /**
   * Get the display name for this object.
   * @abstract
   * @returns {string} Display name
   */
  getDisplayName() {
    throw new Error('Method getDisplayName() must be implemented');
  }
}

/**
 * Represents a user in the system.
 * @extends Entity
 * @implements Named
 */
class User extends Entity {
  /**
   * Create a new user.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   * @param {string} email - Email address
   * @param {string} [role="standard"] - User role
   */
  constructor(id, name, email, role = "standard") {
    super(id, name);
    this.email = email;
    this.role = role;
    this.permissions = new Set();
  }
  
  /**
   * Add a permission to this user.
   * @param {string} permission - Permission to add
   */
  addPermission(permission) {
    this.permissions.add(permission);
  }
  
  /**
   * Check if this user has the specified permission.
   * @param {string} permission - Permission to check
   * @returns {boolean} True if user has the permission
   */
  hasPermission(permission) {
    return this.permissions.has(permission);
  }
  
  /**
   * Get the display name for this user.
   * @returns {string} Display name
   */
  getDisplayName() {
    return `${this.name} (${this.email})`;
  }
}

/**
 * Represents a resource in the system.
 * @extends Entity
 * @implements Named
 * @implements Printable
 */
class Resource extends Entity {
  /**
   * Create a new resource.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   * @param {string} ownerId - Owner's ID
   * @param {string} resourceType - Type of resource
   */
  constructor(id, name, ownerId, resourceType) {
    super(id, name);
    this.ownerId = ownerId;
    this.resourceType = resourceType;
    this.data = {};
  }
  
  /**
   * Update a data field for this resource.
   * @param {string} key - Field key
   * @param {*} value - Field value
   */
  updateData(key, value) {
    this.data[key] = value;
  }
  
  /**
   * Get a data field from this resource.
   * @param {string} key - Field key
   * @param {*} [defaultValue=null] - Default value if key doesn't exist
   * @returns {*} Field value or default
   */
  getData(key, defaultValue = null) {
    return key in this.data ? this.data[key] : defaultValue;
  }
  
  /**
   * Get the display name for this resource.
   * @returns {string} Display name
   */
  getDisplayName() {
    return `${this.name} (${this.resourceType})`;
  }
  
  /**
   * Convert to a printable format.
   * @returns {string} Printable representation
   */
  toPrintableFormat() {
    return `Resource: ${this.name}\nType: ${this.resourceType}\nOwner: ${this.ownerId}\nData: ${JSON.stringify(this.data)}`;
  }
}

/**
 * Represents a document resource.
 * @extends Resource
 */
class Document extends Resource {
  /**
   * Create a new document.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   * @param {string} ownerId - Owner's ID
   * @param {string} [content=""] - Document content
   */
  constructor(id, name, ownerId, content = "") {
    super(id, name, ownerId, "document");
    this.content = content;
    this.version = 1;
  }
  
  /**
   * Update the document content.
   * @param {string} newContent - New content
   */
  updateContent(newContent) {
    this.content = newContent;
    this.version++;
  }
  
  /**
   * Convert to a printable format.
   * @returns {string} Printable representation
   */
  toPrintableFormat() {
    const baseFormat = super.toPrintableFormat();
    return `${baseFormat}\nVersion: ${this.version}\nContent: ${this.content.substring(0, 50)}...`;
  }
}

/**
 * Represents an image resource.
 * @extends Resource
 */
class Image extends Resource {
  /**
   * Create a new image.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   * @param {string} ownerId - Owner's ID
   * @param {number} width - Image width in pixels
   * @param {number} height - Image height in pixels
   * @param {string} [format="png"] - Image format
   */
  constructor(id, name, ownerId, width, height, format = "png") {
    super(id, name, ownerId, "image");
    this.width = width;
    this.height = height;
    this.format = format;
  }
  
  /**
   * Convert to a printable format.
   * @returns {string} Printable representation
   */
  toPrintableFormat() {
    const baseFormat = super.toPrintableFormat();
    return `${baseFormat}\nDimensions: ${this.width}x${this.height}\nFormat: ${this.format}`;
  }
}

/**
 * Represents an admin user with extended permissions.
 * @extends User
 */
class AdminUser extends User {
  /**
   * Create a new admin user.
   * @param {string} id - Unique identifier
   * @param {string} name - Display name
   * @param {string} email - Email address
   */
  constructor(id, name, email) {
    super(id, name, email, "admin");
    // Default admin permissions
    this.addPermission("create_user");
    this.addPermission("delete_user");
    this.addPermission("edit_resource");
  }
  
  /**
   * Check if this admin can manage users.
   * @returns {boolean} True if admin can manage users
   */
  canManageUsers() {
    return this.hasPermission("create_user") && this.hasPermission("delete_user");
  }
  
  /**
   * Get the display name for this admin.
   * @returns {string} Display name
   */
  getDisplayName() {
    return `Admin: ${this.name} (${this.email})`;
  }
}

// Demo usage
if (require.main === module) {
  // Create an admin user
  const admin = new AdminUser("admin1", "Admin User", "admin@example.com");
  admin.addTag("administrator");
  
  // Create a standard user
  const user = new User("user1", "Standard User", "user@example.com");
  user.addTag("beta-tester");
  
  // Create a document
  const doc = new Document("doc1", "Important Document", "user1", "This is an important document");
  doc.addTag("important");
  
  // Create an image
  const img = new Image("img1", "Profile Photo", "user1", 1200, 800);
  img.addTag("profile");
  
  // Demonstrate hierarchy
  const entities = [admin, user, doc, img];
  for (const entity of entities) {
    console.log(`\n${entity}`);
    
    if (entity instanceof User || entity instanceof Resource) { // Both implement Named
      console.log(`Display name: ${entity.getDisplayName()}`);
    }
    
    if (entity instanceof Resource) { // Resources implement Printable
      console.log(`Printable format:\n${entity.toPrintableFormat()}`);
    }
    
    console.log(`Tags: ${[...entity.tags].join(', ')}`);
  }
}

module.exports = {
  Entity,
  Printable,
  Named,
  User,
  Resource,
  Document,
  Image,
  AdminUser
}; 